src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfBuilder.cs

```diff
@@ -0,0 +1,485 @@
+            _dwarfTypes.Add(new DwarfEnumTypeInfo(typeDescriptor, typeRecords, byteSize));
+            return (uint)_dwarfTypes.Count;
+        }
+
+        public uint GetClassTypeIndex(ClassTypeDescriptor classDescriptor)
+        {
+            _dwarfTypes.Add(new DwarfClassTypeInfo(classDescriptor));
+            return (uint)_dwarfTypes.Count;
+        }
+
+        public uint GetCompleteClassTypeIndex(
+            ClassTypeDescriptor classTypeDescriptor,
+            ClassFieldsTypeDescriptor classFieldsTypeDescriptor,
+            DataFieldDescriptor[] fields,
+            StaticDataFieldDescriptor[] statics)
+        {
+            var classInfo = new DwarfClassTypeInfo(classTypeDescriptor, classFieldsTypeDescriptor, fields, statics);
+            _dwarfTypes.Add(classInfo);
+            _staticFields.AddRange(classInfo.StaticVariables);
+            return (uint)_dwarfTypes.Count;
+        }
+
+        public uint GetMemberFunctionTypeIndex(MemberFunctionTypeDescriptor memberDescriptor, uint[] argumentTypes)
+        {
+            _memberFunctionTypeInfos.Add(new MemberFunctionTypeInfo(
+                memberDescriptor,
+                argumentTypes,
+                memberDescriptor.TypeIndexOfThisPointer == GetPrimitiveTypeIndex(TypeFlags.Void)));
+
+            return (uint)_memberFunctionTypeInfos.Count;
+        }
+
+        public uint GetMemberFunctionId(MemberFunctionIdTypeDescriptor memberIdDescriptor)
+        {
+            MemberFunctionTypeInfo memberFunctionTypeInfo = _memberFunctionTypeInfos[(int)memberIdDescriptor.MemberFunction - 1];
+            DwarfClassTypeInfo parentClass = (DwarfClassTypeInfo)_dwarfTypes[(int)(memberIdDescriptor.ParentClass - 1)];
+            DwarfMemberFunction memberFunction = new DwarfMemberFunction(
+                memberIdDescriptor.Name,
+                memberFunctionTypeInfo.MemberDescriptor,
+                memberFunctionTypeInfo.ArgumentTypes,
+                memberFunctionTypeInfo.IsStatic);
+
+            parentClass.AddMemberFunction(memberFunction);
+            _memberFunctions.Add(memberFunction);
+
+            return (uint)_memberFunctions.Count;
+        }
+
+        public string GetMangledName(TypeDesc type)
+        {
+            return _nameMangler.GetMangledTypeName(type);
+        }
+
+        public void EmitSubprogramInfo(
+            string methodName,
+            string sectionSymbolName,
+            long methodAddress,
+            int methodPCLength,
+            uint methodTypeIndex,
+            IEnumerable<(DebugVarInfoMetadata, uint)> debugVars,
+            DebugEHClauseInfo[] debugEHClauseInfos)
+        {
+            if (methodTypeIndex == 0)
+            {
+                return;
+            }
+
+            DwarfMemberFunction memberFunction = _memberFunctions[(int)methodTypeIndex - 1];
+            memberFunction.LinkageName = methodName;
```

Comment:
`LinkageName` vs. `Name` mismatch?

Reply:
This is intentional. Ref to old code: https://github.com/dotnet/llvm-project/blob/e80993c0649d5e949eb4ac23fe22290b4ecb7e66/llvm/tools/objwriter/debugInfo/dwarf/dwarfGen.cpp#L1001

Initially the DWARF tree is built with the debug names for both `Name` and `LinkageName`. The linkage name is then updated here is to the actual mangled name matching the generated code.